### ISSUE TITLE
The pest_deconstruct bikeshed

### ISSUE THREAD

**Issue by CAD97** [pest-parser/pest#317]
[The library.](https://github.com/pest-parser/pest_deconstruct) [Examples.](https://github.com/pest-parser/pest_deconstruct/tree/master/examples) [Experimental version of just the grammar subset of pest_meta using pest_deconstruct.](https://github.com/CAD97/meta-pest-meta)

## So what is `pest_deconstruct`?

I've been experimenting with a way to make working with the `Pairs` API easier. `pest_deconstruct` serves as a middleman between `pest` and your code that helps you turn the "dynamically typed" parse tree into a more "statically typed" syntax tree. It does this via two types: the `FromPest` trait, and the `PestDeconstructor` helper struct.

`pest_deconstruct_derive` is the magic that provides `#[derive(FromPest)]` by utilizing the above helpers. With it, you can write the simple representation of your desired typed syntax tree, and so long as these types line roughly up with the nodes generated by your grammar, just derive the conversion with minimal annotation.

## What's this RFC?

The `FromPest` and `PestDeconstructor` types are likely to be useful for any consumer of the `pest` library, as they are general enough to describe most cases where you want to pull information out of the `Pairs` tree. As such, the RFC is about adding the following API into `pest` proper. However, these names are definitely not finalized, please provide your suggested shed colors. (Documentation comments are truncated for brevity.)

<details><summary>Added API</summary>

### `FromPest`

```rust
pub trait FromPest<'a>: Sized {
    /// The `Rule` enum that defines the parse tree.
    type Rule: pest::RuleType;
    /// The production that this is converted from.
    const RULE: Self::Rule;

    /// Convert from a pest parse tree node.
    /// May panic or return garbage if `pest.as_rule != Self::RULE`.
    fn from_pest(pest: pest::iterators::Pair<'a, Self::Rule>) -> Self;
}

// FUTURE(rust-lang/rust#46969)
impl<'a, T: FromPest<'a>> FromPest<'a> for std::marker::PhantomData<T> {
    type Rule = T::Rule;
    const RULE: T::Rule = T::RULE;
    fn from_pest(pest: Pair<'a, T::Rule>) -> Self {
        let _ = T::from_pest(pest);
        std::marker::PhantomData
    }
}

// FUTURE(rust-lang/rust#46969)
impl<'a, T: FromPest<'a>> FromPest<'a> for Box<T> {
    type Rule = T::Rule;
    const RULE: T::Rule = T::RULE;
    fn from_pest(pest: Pair<'a, T::Rule>) -> Self {
        Box::new(T::from_pest(pest))
    }
}
```

### `PestDeconstructor`

```rust

impl<'i, R: pest::RuleType> pest::iterators::Pair<'i, R> {
    fn deconstruct(self) -> PestDeconstructor<'i, R>;
}
impl<'i, R: pest::RuleType> pest::iterators::Pairs<'i, R> {
    fn deconstruct(self) -> PestDeconstructor<'i, R>;
}

/// Helper for deconstructing a `Pairs` parse tree into its inner productions.
#[derive(Clone, Debug)]
pub struct PestDeconstructor<'i, R: pest::RuleType> { .. }

/// Panic on drop if not fully exhausted to help with correctness
impl<'i, R: pest::RuleType> Drop for PestDeconstructor<'i, R> {
    fn drop(&mut self);
}

impl<'i, R: pest::RuleType> PestDeconstructor<'i, R> {
    /// Convert the next production. For `rule` productions.
    pub fn next<T: FromPest<'i, Rule=R>>(&mut self) -> T;
    /// Convert the next production, if the rule matches. For `rule?` productions.
    pub fn next_opt<T: FromPest<'i, Rule=R>>(&mut self) -> Option<T>;
    /// Convert the next productions, while the rule matches. For `rule*` productions.
    pub fn next_many<T: FromPest<'i, Rule=R>>(&'self mut self) -> impl Iterator<Item=T> + 'self;

    /// Get the next production, without processing it.
    pub fn next_raw(&mut self) -> Option<pest::iterators::Pair<'i, R>>;
    /// Get the next production, if the rule matches, without processing it.
    pub fn next_opt_raw(&mut self, rule: R) -> Option<pest::iterators::Pair<'i, R>>;
    /// Get the next productions, while the rule matches, without processing them.
    pub fn next_many_raw(&'self mut self, rule: R)
        -> impl Iterator<Item=pest::iterators::Pair<'i, R>> + 'self;

    /// Explicitly discard the remaining productions.
    pub fn discard(mut self);
}
```

</details><br>

Not included in this RFC is any of the `pest_deconstruct_derive` machinery, which I would likely rename to `pest-ast` if this RFC is taken. `pest-ast` would continue to exist and evolve outside of `pest`, and could potentially be used by `pest_meta` as well, as it wouldn't introduce any cycles into the dependency graph. (I am uncertain whether `pest_deconstruct` could be used by `pest_meta` as the dependency graph is getting complicated and I haven't traced the whole thing; bootstrap makes everything get confusing.)

## API design notes

- `PestDeconstructor::next_raw` here returns an `Option`, unlike `PestDeconstructor::next`, which asserts that there is a next production to get. It also doesn't take a rule, unlike the other `_raw` fn. `pest_deconstruct` has two separate fn: `next_raw(&mut self, rule: R) -> Pair<'i, R>` and `next_untyped(&mut self) -> Option<Pair<'i, R>>` which have been condensed down into this single fn here in the name of avoiding API bloat. Feel free to argue either direction below; the choice was made arbitrarily.

    Some arguments for `next_raw` just grabbing the next pair (if present):

    - Reduces API duplication with a `next_untyped` or similar
    - More closely represents the raw iteration of `Pair`s

    Some arguments for `next_raw` taking a rule and asserting presence of a pair with that rule:

    - Mirrors `next` more closely
    - More coherent with the other `_raw` fns' formulation

- The `_many` fn return an iterator where `pest_deconstruct` returns a `Vec`. This is an optimization to avoid the `.collect()` allocation and allow the use site to choose how it wants to consume the repetition.

## Open questions

- Should the `_many` iterators require exhaustion like `PestDeconstructor` does? This will require either an explicit `drop(it)` or scoping to get back to being able to use the deconstructor even with NLL, whereas a lack of drop will mean that the lifetime can just be ended by NLL and allow the deconstructor to continue use. I see merit in both allowing `next_many(_raw)` to not be called to exhaustion and in requiring it to be.
- `PestDeconstructor` _could_ return the remaining `pest::iterators::Pairs`, though this could leak implementation details if implementation requires a `Peekable` (thus requiring the return to use `Peekable`), but should it? `next_raw` (or `next_untyped`) allows you to still use it as a raw iterator of `Pair`s.
- Other trait implementations for `PestDeconstructor` and other `Pairs` API that it could also provide.
- All the names (please bikeshed)!
- Others?

## Alternatives

Basically, the status quo: I take the bikeshed from this RFC and apply it to `pest_deconstruct`. (It'll still probably be renamed to `pest-ast`.) We advertise `pest-ast` alongside `pest` as an optional extension (or don't).

Also upstream parts of `pest_deconstruct_derive` into `pest_derive`. This isn't ideal, as these parts of the API, though the most useful parts, are much less developed and have more moving parts to iron out. Additionally, this would probably preclude them being used in `pest_meta`.

**Comment by felix91gr** [comment]
PestPiping, PestWiring?

PestASTer? (A es teer)

**Comment by CAD97** [comment]
The `Pest` prefix to the type name isn't required (rather, it's discouraged) if it's in the `pest` crate.

**Comment by CAD97** [comment]
Some concrete questions raised by a conversation with @vi on Gitter:

- What are the failure cases for `FromPest`? How should they be handled?
  - AST mismatch: this is a programming error and should be a panic. `FromPest<RULE=variant>` should be able to handle any variant of successful parse of that rule.
  - Concrete typing mismatch:

      ```rust
      #[derive(FromPest)]
      #[pest(rule = Rule::integer)]
      struct Integer {
          this: u64,
      }
      ```

     Is this valid? I would tend to say not; I see `FromPest` as roughly `From<pest::iterators::Pair>` with lifetimes allowing the output to borrow from the pest tree. Again, a `FromPest` impl should be a strongly typed version of the parse tree, without adding extra requirements on top.

**Comment by CAD97** [comment]
Some concerns have been raised in the past couple days about the design of `pest_deconstruct`. I'm re-reviewing design decisions and considering changing some decisions, so am postponing this RFC for now.
